File: GAMEOFLIFEJAVA/gameoflife/organisms/plants/SosnowskyHogweed.java
package gameoflife.organisms.plants;

import gameoflife.structs.Vector;
import gameoflife.organisms.Organism;
import gameoflife.organisms.animals.Animal;

public class SosnowskyHogweed extends Plant {

    public SosnowskyHogweed(Vector position){
        super("$", 10, position);
    }

    @Override
    public void Action(){

        Vector[] poisonables = {
            new Vector(0, 1),
            new Vector(1, 1),
            new Vector(1, 0),
            new Vector(1, -1),
            new Vector(0, -1),
            new Vector(-1, -1),
            new Vector(-1, 0),
            new Vector(-1, 1)
        };

        for(Vector poisonable : poisonables){

            Vector position = GetPos().Add(poisonable);

            if(world.OnBoard(position)) {

                if (world.GetOrganism(position) != null) {

                    if (world.GetOrganism(position) instanceof Animal){

                        world.Log(this.getClass().getSimpleName() + " poisoned a " + world.GetOrganism(position).getClass().getSimpleName() + "...");
                        world.GetOrganism(position).Kill();
                    }
                }
            }
        }
        super.Action();
    }

    @Override
    public void Collision(Organism other){

        other.Kill();
        this.Kill();
    }
}

File: GAMEOFLIFE/SosnowskyHogweed.cpp
#include "SosnowskyHogweed.h"
#include "World.h"
#include "Animal.h"

SosnowskyHogweed::SosnowskyHogweed(int x, int y) : Plant('$', 11, x, y, 10, 0) {
}

void SosnowskyHogweed::Action() {

	Plant::Action();

	for (int i = 0; i < 4; i++) {

		auto proximity = world->GetOrganism({ GetX() + moves[i][0], GetY() + moves[i][1]});

		if (proximity != nullptr) {
			bool animal = dynamic_cast<Animal*>(proximity);
			if (animal) {
				world->AddLog(this, "poisoned an animal in its proximity", 2);
				proximity->Kill();
			}
		}
	}
}

void SosnowskyHogweed::Collision(Organism* other) {
	world->AddLog(this, "poisoned an animal and died", 2);
	other->Kill();
	Kill();
}

Organism* SosnowskyHogweed::Child() {
	return new SosnowskyHogweed(*this);
}